src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs

```diff
@@ -986,22 +1144,50 @@ public static unsafe SocketError ReceiveAsync(SafeCloseSocket handle, byte[] buf
 
         public static unsafe SocketError ReceiveAsync(SafeCloseSocket handle, IList<ArraySegment<byte>> buffers, SocketFlags socketFlags, OverlappedAsyncResult asyncResult)
         {
-            // Set up asyncResult for overlapped WSASend.
+            // Set up asyncResult for overlapped WSARecv.
+            // This will pin the buffers.
             asyncResult.SetUnmanagedStructures(buffers);
             try
             {
-                // This can throw ObjectDisposedException.
+                int result;
                 int bytesTransferred;
-                SocketError errorCode = Interop.Winsock.WSARecv(
-                    handle,
-                    asyncResult._wsaBuffers,
-                    asyncResult._wsaBuffers.Length,
-                    out bytesTransferred,
-                    ref socketFlags,
-                    asyncResult.OverlappedHandle,
-                    IntPtr.Zero);
 
-                return asyncResult.ProcessOverlappedResult(errorCode == SocketError.Success, bytesTransferred);
+                int bufferCount = buffers.Count;
+                if (bufferCount <= WSABuffer.StackAllocLimit)
+                {
+                    WSABuffer* wsaBuffers = stackalloc WSABuffer[bufferCount];
+                    PopulateWSABufferArray(wsaBuffers, buffers);
+
+                    // This can throw ObjectDisposedException.
+                    result = Interop.Winsock.WSARecv(
+                        handle,
+                        wsaBuffers,
+                        bufferCount,
+                        &bytesTransferred,
+                        &socketFlags,
+                        asyncResult.OverlappedHandle,
+                        null);
+                }
+                else
+                {
+                    WSABuffer[] wsaHeapBuffers = new WSABuffer[bufferCount];
+                    fixed (WSABuffer* wsaBuffers = wsaHeapBuffers)
+                    {
+                        PopulateWSABufferArray(wsaBuffers, buffers);
+
+                        // This can throw ObjectDisposedException.
```

Comment:
I'm not sure we need all of these "This can throw ODE" comments... any P/Invoke that takes a SafeHandle can throw an ODE if the handle has been disposed.